### PR TITLE
add test to compare actual github source with config-osg repo

### PR DESCRIPTION
**Pull request by zvada**
Pull latest osg branch of config-repo and compare with actual configuration in /cvmfs/config-osg.opensciencegrid.org/. Applies to the `-itb` host(s) only.

### CODE REVIEW

**Review by DrDaveD — CHANGES_REQUESTED** [review]
I had in mind instead to check for differences before doing git pull, to see if there were any manual changes done.  Then if there are no differences, go ahead and do the git pull, rsync and publish.  Also, have a way to override the differences; I was thinking it would prompt and say something like "are you sure", but alternatively you could add a command line option to override the differences, and  remind the user about the option in the error message when there's a difference.  The publishing part could share code, perhaps via a function that takes a parameter of a source directory to rsync from.

**Comment by zvada** [comment]
The workflow we discussed:

```
try something first in /cvmfs/config-osg.opensciencegrid.org on hcc-oasis-itb, then make a pull request to get it into github, then once merged run copy_config_osg to make sure everything matches github, and finally run copy_config_osg on hcc-oasis to put into production.
```

The review I read above changes assignment, however. I wonder why you think code in the review does something different than discussed. My expectation is, if you do manual changes in a first step on the `-itb` in transaction mode, you need to publish it on `-itb` prior to test from the client side. No rsync needed at that stage on the `-itb` host at all. If test is OK, we typically do pull request with approved/tested changes and run `copy_osg_config` on the production host which does copy from `-itb`, rsync and publish.
Still, we didn't want to wait for puppet do git pull on /opt/config-repo and have possibility run manually `copy_osg_config` on `-itb` host that pulls the change and compare differences before puppet runs. Assuming published changes in the `-itb` were intact in between pull request and `copy_osg_config` all is good and one can run `copy_osg_config` on the production host. In that case, I'm not sure about the necessity of override option rather prompt to an error that would say something is different thus operator need to check if PR was submitted and approved because things should be up to date otherwise. We can perhaps make the error more verbal in the step of diff. Or, let's talk about this once again in case I misunderstood the workflow.

**Comment by DrDaveD** [comment]
Ok we discussed on slack, and the bottom line is that I believe you're right and the way I was thinking about it was incorrect.

Instead, I propose these two changes for helping to automate the workflow:

1. Move the verification of what's in the -itb config repository against git to be run on the production machine before publication, and enable it to be bypassed with a -f option.  In this way we'll make sure the verification is done but still allow somebody to not wait for git to catch up.
2. Add new functionality for the -itb machine to assist in making the PR: have it verify that it is being run as a non-root user, that ~/work/config-repo exists, and that it is in the osg branch, and if those are OK then merge any changes from upstream and rsync everything in /cvmfs/config-osg.opensciencegrid.org/* to ~/work/config-repo.  I think it's best to stop at that point and let the user do diffs, optionally create a PR branch, commit, push, and create a PR, but you might disagree and want to automate more of those steps.

**Comment by DrDaveD** [comment]
Alternatively for part 2 it doesn't have to assume that the git repository is in ~/work/config-repo if it is able to verify that the current directory is a checked-out copy of the git repo.

File: goc/bin/copy_config_osg
Comment (DrDaveD) [inline goc/bin/copy_config_osg]: Instead of having both of these in the same "if", such that it fails over to the other behavior when running as root on -itb, only check here for -itb and then inside the case immediately check for non-root and fail if it is root.    Also in the other case, on non-itb, check immediately to make sure running as root.

File: goc/bin/copy_config_osg
Comment (DrDaveD) [inline goc/bin/copy_config_osg]: This will return true inside of any git repo.  I suggest instead to use
```
if git remote -v 2>/dev/null|grep -q /config-repo.git
```

File: goc/bin/copy_config_osg
Comment (DrDaveD) [inline goc/bin/copy_config_osg]: I had in mind to merge from upstream.  Since this git repo is used to make pull requests, the users have to fork it and that should be their origin.  You can use `git pull upstream osg`.  I think it's better not to redirect output to /dev/null, go ahead and let it show.  It probably also ought to exit if the the command fails, for example if there is no upstream remote.

File: goc/bin/copy_config_osg
Comment (DrDaveD) [inline goc/bin/copy_config_osg]: No, here I had in mind to do the rsync from /cvmfs/config-osg.opensciencegrid.org/  to the current directory, so the user could then proceed to make a pull request out of it.

File: goc/bin/copy_config_osg
Comment (DrDaveD) [inline goc/bin/copy_config_osg]: This could just do the git checkout osg instead of failing.

File: goc/bin/copy_config_osg
Comment (DrDaveD) [inline goc/bin/copy_config_osg]: It's OK for this to just be a git pull because there will be only one remote.  Again though I think it would be better to not redirect to /dev/null, let the updates show.

File: goc/bin/copy_config_osg
Comment (DrDaveD) [inline goc/bin/copy_config_osg]: I suggest adding ` [y/N]` to that prompt.

**Review by DrDaveD — CHANGES_REQUESTED** [review]

File: goc/bin/copy_config_osg
Comment (zvada) [reply goc/bin/copy_config_osg]: ok, added

File: goc/bin/copy_config_osg
Comment (zvada) [reply goc/bin/copy_config_osg]: OK

File: goc/bin/copy_config_osg
Comment (zvada) [reply goc/bin/copy_config_osg]: OK, changed.

File: goc/bin/copy_config_osg
Comment (zvada) [reply goc/bin/copy_config_osg]: OK, hope I got it right in `488ccf5`.

File: goc/bin/copy_config_osg
Comment (zvada) [reply goc/bin/copy_config_osg]: OK, changed.

File: goc/bin/copy_config_osg
Comment (zvada) [reply goc/bin/copy_config_osg]: OK, changed.

File: goc/bin/copy_config_osg
Comment (zvada) [reply goc/bin/copy_config_osg]: OK, splitted.

File: goc/bin/copy_config_osg
Comment (DrDaveD) [inline goc/bin/copy_config_osg]: Instead of this whole section I think this should fail with an error message about needing to run as non-root on itb.  If run as root on itb, there's no value in copying the current repo into /opt/config-repo that I can see.  Nobody will make a pull request from there.

File: goc/bin/copy_config_osg
Comment (DrDaveD) [inline goc/bin/copy_config_osg]: This part needs to confirm that it is in the osg branch before doing the git pull upstream.

Comment (DrDaveD) [reply goc/bin/copy_config_osg]: Also have the script fail if the git pull upstream osg fails.  In fact, go ahead and put set -ex just before the 'cd' above; that will make it more verbose and exit if there are any command failures.

File: goc/bin/copy_config_osg
Comment (DrDaveD) [inline goc/bin/copy_config_osg:75]: Although this is unlikely to fail, it would be better to do this check first, before doing all that stuff that it later cleans up (the making of a temporary config file, mounting a repo, and starting a transaction).  It can be first after the set -ex above.

**Review by DrDaveD — CHANGES_REQUESTED** [review]

File: goc/bin/copy_config_osg
Comment (zvada) [reply goc/bin/copy_config_osg]: so we're limiting to run this only as non-root? I thought we want to have an option let this run as root on -itb host like on the production host. But perhaps if we leave option for root on -itb we shouldn't run rsync but only throw an error that current pull in /opt/config-repo doesn't have latest changes yet compared to /cvmfs/config-osg.opensciencegrid.org/ on -itb host. What do you think?

Comment (DrDaveD) [reply goc/bin/copy_config_osg]: I don't think it's useful to run this as root on itb, because we are doing that same check now on production in order to simplify.  Just limit it to run as non-root.

File: goc/bin/copy_config_osg
Comment (zvada) [reply goc/bin/copy_config_osg:75]: ok, sounds good.

File: goc/bin/copy_config_osg
Comment (zvada) [reply goc/bin/copy_config_osg]: OK.

File: goc/bin/copy_config_osg
Comment (zvada) [reply goc/bin/copy_config_osg]: ok, but we still modify the repo as root on -itb, because one needs to run transaction, then edit things, publish ... If we leave out out the whole `else` section and then throw an error useage allowed for non-root only, we probably have to document for the operator that it's necessary to switch to non-root to use `copy_config_osg` after the tests of changes in the /cvmfs/config-osg.opensciencegrid.org/ on -itb are working. Is this how you want it?

Comment (DrDaveD) [reply goc/bin/copy_config_osg]: Yes

Comment (zvada) [reply goc/bin/copy_config_osg]: OK.

File: goc/bin/copy_config_osg
Comment (DrDaveD) [reply goc/bin/copy_config_osg]: Getting close.  Still waiting on the confirmation that is in the osg branch.

File: goc/bin/copy_config_osg
Comment (DrDaveD) [inline goc/bin/copy_config_osg:75]: I'm sorry, I should have said to put this just before the set -ex instead of just after it.  It's very minor, but might as well do it the better way while we're in here and there's still something more substantial left to change.

**Review by DrDaveD — CHANGES_REQUESTED** [review]

File: goc/bin/copy_config_osg
Comment (zvada) [reply goc/bin/copy_config_osg]: didn't you say should it fail if it's not on osg branch? that's why `set -ex` was added and `git pull upstream osg` will fail if it's not on osg branch and exit, right? ... or did you want yet another `if [[ ${current_branch_name} == "osg" ]]; then...` before `git pull upstream osg` followed up by `rsync...`?

**Comment by DrDaveD** [comment]
After approving and merging my other PR #120, merge upstream/master into your branch and update goc/rpm/oasis-server.spec for a new release, and include that in this PR too.

File: goc/bin/copy_config_osg
Comment (zvada) [reply goc/bin/copy_config_osg:75]: OK.

File: goc/bin/copy_config_osg
Comment (DrDaveD) [reply goc/bin/copy_config_osg]: `git pull upstream osg` will not fail if the current working branch is not osg, it will just merge the upstream osg branch into whatever the current branch is.  Yes there should be a check for current_branch_name == osg, else fail.  (After that the developer should normally create a new branch before committing the changes and making a PR).

Comment (zvada) [reply goc/bin/copy_config_osg]: OK, makes sense now, wasn't sure how it behaves, should have tested myself and see, though.

File: goc/bin/copy_config_osg
Comment (DrDaveD) [inline goc/bin/copy_config_osg]: Now I just see a typo left: use "switch" instead of "swith".

**Review by DrDaveD — CHANGES_REQUESTED** [review]
And update oasis-goc.spec for a new release.

**Review by DrDaveD — APPROVED** [review]

**Pull request merged**
